crates/next-dev/src/lib.rs

```diff
@@ -143,29 +150,75 @@ impl NextDevServerBuilder {
         let console_ui = Arc::new(ConsoleUi::new(log_options));
         let console_ui_to_dev_server = console_ui.clone();
 
-        let server = DevServer::listen(
-            turbo_tasks.clone(),
-            move || {
-                source(
-                    root_dir.clone(),
-                    project_dir.clone(),
-                    entry_requests.clone(),
-                    eager_compile,
-                    turbo_tasks.clone().into(),
-                    console_ui.clone().into(),
-                    browserslist_query.clone(),
-                    server_component_externals.clone(),
-                )
-            },
-            (
-                self.hostname.context("hostname must be set")?,
-                self.port.context("port must be set")?,
+        let start_port = self.port.context("port must be set")?;
+        let host = self.hostname.context("hostname must be set")?;
+
+        let mut err: Option<anyhow::Error> = None;
+
+        let tasks = turbo_tasks.clone();
+        let source = move || {
+            source(
+                root_dir.clone(),
+                project_dir.clone(),
+                entry_requests.clone(),
+                eager_compile,
+                turbo_tasks.clone().into(),
+                console_ui.clone().into(),
+                browserslist_query.clone(),
+                server_component_externals.clone(),
             )
-                .into(),
-            console_ui_to_dev_server,
-        );
+        };
+
+        // Retry to listen on the different port if the port is already in use.
+        for retry_count in 0..10 {
+            let current_port = start_port + retry_count;
+            let addr = SocketAddr::new(host, current_port);
+
+            let listen_result = DevServer::listen(
+                tasks.clone(),
+                source.clone(),
+                addr,
+                console_ui_to_dev_server.clone(),
+            );
+
+            match listen_result {
+                Ok(server) => {
+                    return Ok(server);
+                }
+                Err(e) => {
+                    let should_retry = if self.allow_retry {
+                        e.source()
+                            .map(|e| {
+                                e.source()
+                                    .map(|e| {
+                                        e.downcast_ref::<std::io::Error>()
+                                            .map(|e| e.kind() == std::io::ErrorKind::AddrInUse)
+                                            == Some(true)
+                                    })
+                                    .unwrap_or_else(|| false)
+                            })
+                            .unwrap_or_else(|| false)
```

Comment:
Is this because we used `err.context(...)`? Please add a comment.

Reply:
yes, it's wrapped with anyhow:Error. I tried to avoid wrap it internally, then now error type becomes `hyper::Error` which `next-dev` does not imports it directly - so choose to unwrap instead of brining hyper as direct dep in next-dev.